src/entity/components/Animatable.ts

```diff
@@ -0,0 +1,69 @@
+import type { ECSComponent } from '@/ecs/ECSComponent';
+import type { ECSEntity } from '@/ecs/ECSEntity';
+import type { SpriteName } from '@/renderer/createAnimatedSprite';
+import type { Nullable, Values } from '@/utils/types';
+
+export const AnimationState = {
+  IDLE: 'idle',
+  RUNNING: 'running',
+  ATTACKING: 'attacking'
+} as const;
+export type AnimationState = Values<typeof AnimationState>;
+
+export type AnimationOptions = {
+  loop: boolean;
+  animationSpeed: number;
+  fallbackOnComplete: Nullable<AnimationState>;
+};
+
+export const AnimatableBrand = 'animatable';
+type AnimatableBrand = typeof AnimatableBrand;
+
+export type Animatable = ECSComponent<
+  AnimatableBrand,
+  {
+    state: AnimationState;
+    spriteName: SpriteName;
+    isDirty: boolean;
+    options: AnimationOptions;
+  }
+>;
+
+export function hasAnimatable<E extends ECSEntity>(e: E): e is E & Animatable {
+  return AnimatableBrand in e;
+}
+
+export function animatableComponent(spriteName: SpriteName): Animatable {
+  let _state: AnimationState = AnimationState.IDLE;
+  // We need to track the dirtyness of the new animation state, otherwise the animation system will update and reset the texture on every tick
+  let _isDirty = false;
+
```

Comment:
This will most likely break under serialization. We could have tagging utilities that apply code to components with certain tags, but a component itself should not be stateful / reactive logic.

We could consider something like
```
// def
world::registerDecorator<C: ECSComponent>(brand: C[brand], decorate: (comp: C) => C)

// usage
// decorate is a function that applies a proxy and does the dirty tracking
world.registerDecorator<Animatable>('animateable', (comp) => decorate(comp))
```
Over time, we could add a few "standard" decorators, like dirty.

Alternatively, the system responsible for setting the new state should also update the dirty flag.